### PR TITLE
mesa: update to 20.0.0, HGL fixes

### PR DESCRIPTION
**Pull request by X547**
BGLView::CopyPixelsIn/Out is not yet implemented.

Front and back buffer BBitmap are now allocated in winsys, memory copy count is
reduced.

Fixes:
	#4471
	#4466
	Fix buffer leak.

New features:
	Double buffering.
	Shared contexts (needed for Blender).

### CODE REVIEW

File: sys-libs/mesa/mesa-20.0.0.recipe [hunk 1]
```diff
@@ -0,0 +1,160 @@
+SUMMARY="Multi-platform GL implementation"
+DESCRIPTION="Mesa is an open-source implementation of the OpenGL \
+specification. The OpenGL specification documents a system for rendering \
+interactive 3D graphics. Mesa fills the role of the Haiku OpenGL kit \
+providing 3D rendering to Haiku applications."
+HOMEPAGE="https://www.mesa3d.org/"
+COPYRIGHT="1999-2018 Brian Paul"
```
Comment (diversys) [inline sys-libs/mesa/mesa-20.0.0.recipe]: Copyright years should be updated i guess

**Comment by diversys** [comment]
Doesn't compile here on hrev53898 x86_64

```
In file included from ../src/mesa/main/context.c:143:
src/mesa/main/dispatch.h:8116:10: warning: null character(s) ignored
 static inline void SET_IsQuery(struct _glapi_table *disp, GLboolean (GLAPIENTRYP fn)(GLuint)) {
          ^
src/mesa/main/dispatch.h:8116:8: error: unknown type name 'in'
 static inline void SET_IsQuery(struct _glapi_table *disp, GLboolean (GLAPIENTRYP fn)(GLuint)) {
        ^~
src/mesa/main/dispatch.h:8116: error: expected '=', ',', ';', 'asm' or '__attribute__' before ')' token
 static inline void SET_IsQuery(struct _glapi_table *disp, GLboolean (GLAPIENTRYP fn)(GLuint)) {
 
src/mesa/main/dispatch.h:8117:1: error: expected identifier or '(' before '}' token
    SET_by_offset(disp, _gloffset_IsQuery, fn);
 ^
../src/mesa/main/context.c: In function 'create_beginend_table':
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsProgram'; did you mean 'SET_Histogram'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1107:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsProgram);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_IsProgram'; did you mean 'GET_Histogram'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1107:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsProgram);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsVertexArray'; did you mean 'glIsVertexArray'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1108:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsVertexArray);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_IsVertexArray'; did you mean 'glIsVertexArray'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1108:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsVertexArray);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsEnabledi'; did you mean 'SET_IsEnabled'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1111:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsEnabledi);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_IsEnabledi'; did you mean 'GET_IsEnabled'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1111:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsEnabledi);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsRenderbuffer'; did you mean 'glIsRenderbuffer'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1112:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsRenderbuffer);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_IsRenderbuffer'; did you mean 'glIsRenderbuffer'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1112:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsRenderbuffer);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsFramebuffer'; did you mean 'glIsFramebuffer'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1113:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsFramebuffer);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_IsFramebuffer'; did you mean 'glIsFramebuffer'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1113:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsFramebuffer);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_CheckFramebufferStatus'; did you mean 'glCheckFramebufferStatus'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1114:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(CheckFramebufferStatus);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_CheckFramebufferStatus'; did you mean 'glCheckFramebufferStatus'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1114:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(CheckFramebufferStatus);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_GetStringi'; did you mean 'SET_GetString'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1117:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(GetStringi);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_GetStringi'; did you mean 'GET_GetString'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1117:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(GetStringi);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsQuery'; did you mean 'GET_IsQuery'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1119:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsQuery);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsSampler'; did you mean 'glIsSampler'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1120:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsSampler);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_IsSampler'; did you mean 'glIsSampler'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1120:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsSampler);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsSync'; did you mean 'SET_IsList'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1121:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsSync);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_IsSync'; did you mean 'GET_IsList'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1121:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsSync);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_IsTransformFeedback'; did you mean 'glIsTransformFeedback'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1123:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsTransformFeedback);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_IsTransformFeedback'; did you mean 'glIsTransformFeedback'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1123:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(IsTransformFeedback);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_FenceSync'; did you mean 'glFenceSync'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1126:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(FenceSync);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_FenceSync'; did you mean 'glFenceSync'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1126:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(FenceSync);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_ClientWaitSync'; did you mean 'glClientWaitSync'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1127:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(ClientWaitSync);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_ClientWaitSync'; did you mean 'glClientWaitSync'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1127:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(ClientWaitSync);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_MapBuffer'; did you mean 'SET_IsBuffe'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1128:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(MapBuffer);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_MapBuffer'; did you mean 'GET_IsBuffe'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1128:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(MapBuffer);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_UnmapBuffer'; did you mean 'SET_ReadBuffer'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1129:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(UnmapBuffer);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_UnmapBuffer'; did you mean 'GET_ReadBuffer'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1129:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(UnmapBuffer);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_MapBufferRange'; did you mean 'glMapBufferRange'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1130:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(MapBufferRange);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_MapBufferRange'; did you mean 'glMapBufferRange'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1130:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(MapBufferRange);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_ObjectPurgeableAPPLE'; did you mean 'glObjectPurgeableAPPLE'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1131:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(ObjectPurgeableAPPLE);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_ObjectPurgeableAPPLE'; did you mean 'glObjectPurgeableAPPLE'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1131:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(ObjectPurgeableAPPLE);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:29: error: implicit declaration of function 'SET_ObjectUnpurgeableAPPLE'; did you mean 'glObjectUnpurgeableAPPLE'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                             ^~~~
../src/mesa/main/context.c:1132:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(ObjectUnpurgeableAPPLE);
    ^~~~~~~~~~~~~
../src/mesa/main/context.c:1104:47: error: implicit declaration of function 'GET_ObjectUnpurgeableAPPLE'; did you mean 'glObjectUnpurgeableAPPLE'? [-Werror=implicit-function-declaration]
 #define COPY_DISPATCH(func) SET_##func(table, GET_##func(ctx->Exec))
                                               ^~~~
../src/mesa/main/context.c:1132:4: note: in expansion of macro 'COPY_DISPATCH'
    COPY_DISPATCH(ObjectUnpurgeableAPPLE);
    ^~~~~~~~~~~~~
cc1: some warnings being treated as errors
[424/942] Compiling C object 'src/mesa/bbe4a73@@mesa_commo...son-generated_.._.._mapi_glapi_gen_marshal_generated.c.o'
ninja: build stopped: subcommand failed.
Warning: Command '['bash', '-c', '. /wrapper-script']' returned non-zero exit status 1

```

**Comment by X547** [comment]
> src/mesa/main/dispatch.h:8116:10: warning: null character(s) ignored
Strange, look like file corruption.

**Comment by X547** [comment]
Regression: Blender 2.81 crash when opening material preview, work fine with mesa 19.3.3.

**Comment by waddlesplash** [comment]
Re EGL: yes, I think it was broken, unfortunately, and I never got around to looking into that for some apps I was trying to port. There are some examples around somewhere for the EGL API if you want to look into that.

**Comment by X547** [comment]
Added EGL test application: https://github.com/X547/HaikuUtils/tree/master/EGLTest. Currently crash in egl_haiku.cpp: haiku_make_current.

**Comment by diversys** [comment]
> > src/mesa/main/dispatch.h:8116:10: warning: null character(s) ignored
> 
> Strange, look like file corruption.

Indeed, tried running checkfs (which didn't reveal any errors) and then building mesa succeeded.

**Comment by diversys** [comment]
Haiku3d is still black and white tho with current mesa 20 on x86_64.

**Comment by X547** [comment]
> Haiku3d is still black and white tho with current mesa 20 on x86_64.

It look like unrelated to HGL, it is Mesa core or Haiku3d problem.

**Comment by kenmays** [comment]
MesaDemos has EGL programs like: eglinfo, peglgears

eglinfo
> EGL client extensions string:
>     EGL_EXT_client_extensions EGL_EXT_platform_base
>     EGL_KHR_client_get_all_proc_addresses EGL_KHR_debug

peglgears
> peglgears: EGL version = 1.4
> peglgears: EGL_VENDOR = Mesa Project

mesa:egl/egl-symbols-check fails unknown Haiku-related exported symbols. Same for Shared GL API (GL-ES)

egl-symbols-check
> src/egl/libEGL.so.1.0.0: unknown symbol exported: __bss_start
> src/egl/libEGL.so.1.0.0: unknown symbol exported: _edata
> src/egl/libEGL.so.1.0.0: unknown symbol exported: _eglInitDriver
> src/egl/libEGL.so.1.0.0: unknown symbol exported: _end
> src/egl/libEGL.so.1.0.0: unknown symbol exported: _fini
> src/egl/libEGL.so.1.0.0: unknown symbol exported: _gSharedObjectHaikuABI
> src/egl/libEGL.so.1.0.0: unknown symbol exported: _gSharedObjectHaikuVersion
> src/egl/libEGL.so.1.0.0: unknown symbol exported: _init
> src/egl/libEGL.so.1.0.0: unknown symbol exported: haiku_create_context
> src/egl/libEGL.so.1.0.0: unknown symbol exported: haiku_destroy_context
> src/egl/libEGL.so.1.0.0: unknown symbol exported: haiku_make_current
> src/egl/libEGL.so.1.0.0: unknown symbol exported: haiku_swap_buffers
> src/egl/libEGL.so.1.0.0: unknown symbol exported: haiku_terminate
> src/egl/libEGL.so.1.0.0: unknown symbol exported: init_haiku

**Comment by kenmays** [comment]
Haiku3D: https://github.com/haiku/haiku/blob/master/src/apps/haiku3d/texture/VideoFileTexture.cpp

```
			// Create Texture
			glGenTextures(1, &fId);
			glBindTexture(GL_TEXTURE_2D, fId);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
			glTexImage2D(GL_TEXTURE_2D, 0, 4,
				(int) fVideoBitmap->Bounds().Width() + 1,
				(int) fVideoBitmap->Bounds().Height() + 1,
				0, GL_BGRA, GL_UNSIGNED_BYTE, fVideoBitmap->Bits());
```

**Comment by X547** [comment]
No crash for Blender 2.82 when opening material preview. In Blender 2.81 crash appear in generated LLVM code for shader (lp_rast.c:lp_rast_shade_quads_mask).

**Comment by X547** [comment]
Haiku3d is fixed by https://review.haiku-os.org/c/haiku/+/2269.

**Comment by kenmays** [comment]
NOTE: I isolated two 'timeout issues' with Mesa 20.0.0 versus Mesa 19.3.4:

src/compiler/glsl/tests/cache_test
```
Failed to create ./cache-test-tmp/xdg-cache-home for shader cache (No such file or directory)---disabling.
Failed to create ./cache-test-tmp/mesa-glsl-cache-dir for shader cache (No such file or directory)---disabling.
```

src/gallium/targets/osmesa/osmesa-render
```
Running main() from gtest_main.cc
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from OSMesaRenderTest/OSMesaRenderTestFixture
[ RUN      ] OSMesaRenderTest/OSMesaRenderTestFixture.Render/rgba_unsigned_byte
```

**Comment by kenmays** [comment]
@X547 - Mesa 20.0.1 stable release is due around May 4th. We could use Mesa 19.3.4 / Blender 2.82  until the newer Mesa-20.0.x/LLVM bugs are sorted out.

**Comment by X547** [comment]
Crash bug was fixed in upstream, I included patch. Now mesa-19.3.4 can be abandoned.

**Comment by kenmays** [comment]
Verified Mesa 20.0.0 - with today's updated patch, all unit tests: PASS. We can officially decommission Mesa 19.3.x.

**Comment by kenmays** [comment]
@X547 - Can you review?

glteapot:

https://github.com/haiku/haiku/blob/master/src/apps/glteapot/ObjectView.cpp
ObjectView::ObjectAtPoint(const BPoint &point)
```
	glReadBuffer(GL_BACK);
	uchar pixel[256];
	glReadPixels((GLint)point.x, (GLint)(Bounds().bottom - point.y), 1, 1,
		GL_RGB, GL_UNSIGNED_BYTE, pixel);
	int objNum = pixel[0];
	objNum = (255 - objNum) >> 3;
```

ObjectView::SpinIt()
```
if (fFog != fLastFog) {
		if (fFog) {
			glFogf(GL_FOG_START, 1.0);
			glFogf(GL_FOG_DENSITY, 0.2);
			glFogf(GL_FOG_END, depthOfView);
			glFogfv(GL_FOG_COLOR, foggy);
			glEnable(GL_FOG);
			bgColor = foggy;
			glClearColor(bgColor[0], bgColor[1], bgColor[2], 1.0);
		} else {
			glDisable(GL_FOG);
			bgColor = black;
			glClearColor(bgColor[0], bgColor[1], bgColor[2], 1.0);
		}
		fLastFog = fFog;
		changed = true;
```
NOTE: LockGL/UnLockGL - inconsistency in gl lock wrapping. Seems to affect glReadPixels / mouse control of spinning teapots. BGLView-fix related?

**Comment by waddlesplash** [comment]
Can you submit this upstream (to https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests )?

@kallisti5 has push access there, and can review/+2 it.

**Comment by X547** [comment]
Note that upstream build is broken on Haiku. There are internal mesa function `int64` that collides with Haiku `int64`.

**Comment by CodeforEvolution** [comment]
Good evening! I was wondering if there was anything I could do to help get this merged, as I'm currently working on updating the Godot game engine port, which requires OpenGL to draw its entire interface.

**Comment by X547** [comment]
You can use this patch locally for now. Unimplemented features should not affect Godot.

> I'm currently working on updating the Godot game engine port

Then please remove BDirectWindow, it is broken and can't be fixed by design. It also don't give significant performance improvements.

**Comment by X547** [comment]
Current HaikuPorts Godot version 2.1.6-2 crashes at exit, because it attempts to delete BGLView. Attached views are deleted automatically when window is closed and deleted.

**Comment by CodeforEvolution** [comment]
Thank you for the insight! I'll most likely replace the BDirectWindow with a BWindowScreen.

**Comment by X547** [comment]
> I'll most likely replace the BDirectWindow with a BWindowScreen.

You should not use direct framebuffer access in windowed mode, otherwise cursor painting will be broken. Regular BWindow is enough.

**Comment by CodeforEvolution** [comment]
Alright, I will just utilize a BWindow.

**Comment by waddlesplash** [comment]
@X547: Any chance you can work on CopyPixels and then upstream this to Mesa trunk? It would be nice to get our Mesa situation solidified before embarking on DRM or something like that...

**Comment by diversys** [comment]
Can we merge it now?

**Comment by X547** [comment]
It currently fails to build because of [hrev54701](https://git.haiku-os.org/haiku/commit/?h=hrev54701&id=b05d6f0af0d5ba3f18462ceb1929bb4ab6e0c03a). Mesa expects `asprintf` to be declared. Using BSD headers will cause conflicts with internal Mesa declarations. Also newer Mesa versions cause conflict with int32 from `SupportDecls.h`.

**Comment by X547** [comment]
I managed to fix build with inserting following code to all sources that use `asprintf`, `vasprintf`:
```
#ifdef __HAIKU__
int asprintf(char **ret, char const *format, ...) __PRINTFLIKE(2,3);
int vasprintf(char **ret, char const *format, va_list ap);
#endif
```

**Comment by waddlesplash** [comment]
@kallisti5 should review this and the new APIs at least before it is merged.

**Comment by kallisti5** [comment]
Hi!  We fixed the asprintf issue [here](https://git.haiku-os.org/haiku/commit/?id=db7dcab7b16610ecdcbc6b670fad6b36703435fa)

I have a few basic patches for Haiku support here waiting on CR:
  https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/8266

Once I get these basics merged upstream, i'll work on getting your patches merged as well.

**Comment by kallisti5** [comment]
@X547 Just a heads up i'm working through these patches upstream here (also rebasing, fixing some minor stuff, and doing some cleanup)

https://gitlab.freedesktop.org/kallisti5/mesa/-/commits/jan2021-haiku-fixes

**Comment by kenmays** [comment]
@kallisti5 - Mesa 20.0.8 builds fine on Haiku R1B2 (I retested the build  today).

**Comment by kenmays** [comment]
@kallisti5 and @X547 - Got Mesa 20.3.2 (git) successfully compiled today on Haiku R1B2. Thank you.

**Comment by X547** [comment]
Merged upstream in https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/8323.

**Comment by waddlesplash** [comment]
We should open a new PR with the updated recipe, right?

**Comment by X547** [comment]
Yes, but not here.

**Pull request closed without merging**